File: Lab4/fork_and_wait.py
import os
import time

def main():
    pid = os.fork()

    if pid < 0:
        raise Exception("Fork failed")
    elif pid == 0:
        # Child process
        print(f"Child process {os.getpid()} is running")
        time.sleep(2)
        print(f"Child process {os.getpid()} is exiting")
    else:
        # Parent process
        print(f"Parent process {os.getpid()} is waiting for child process {pid}")
        os.waitpid(pid, 0)
        print(f"Parent process {os.getpid()} resumes execution")

if __name__ == "__main__":
    main()

File: Lab4/fork_and_wait.c
#include <stdio.h>
#include <unistd.h>
#include <sys/wait.h>

int main() {
    pid_t pid;

    pid = fork();

    if (pid < 0) {
        perror("Fork failed");
        return 1;
    } else if (pid == 0) {
        // Child process
        printf("Child process %d is running\n", getpid());
        sleep(2);
        printf("Child process %d is exiting\n", getpid());
        return 0;
    } else {
        // Parent process
        printf("Parent process %d is waiting for child process %d\n", getpid(), pid);
        wait(NULL);
        printf("Parent process %d resumes execution\n", getpid());
    }

    return 0;
}